[Bug]: `ts-jest` silently hangs up with certain `moduleResolution` options

### Version

29.1.1

### Steps to reproduce

1. Open minimal reproduction https://stackblitz.com/edit/sbnc-tsnode-hang-issue?file=package.json
2. Wait for `npm install`
3. Run `npm run test`

### Expected behavior

The test should either run or give some error.

### Actual behavior

The test hangs up forever.

### Debug log

You can find the debug log in the reproduction: `ts-jest.log`

### Additional context

Please check out `jest.config.js`. It only has two meaningful config items: `extensionsToTreatAsEsm` and `transform.isolatedModules`. Comment out any of them, and the test run fails properly with an error message. Leave them both in, and the test hangs up forever.

The issue is related to `moduleResolution` in `tsconfig.json`. Set it to `node10` or `classic` and there is no hang, set it to `node16`, `nodenext` or `bundler` and the issue comes up.

Please note, that it is very minimal, I've removed everything not related to the issue. So ignore if this configuration does not make sense, I've narrowed it down from a much more complex real issue I had. The main point here is that `ts-jest` hangs without any indication of an error, which is very painful to debug.

The reproduction is on node v16, but the same happens on Node v20 as well.


### Environment

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (4) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 16.20.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.4.2 - /usr/local/bin/npm
    pnpm: 8.6.10 - /usr/local/bin/pnpm
  npmPackages:
    jest: ^29.0.0 => 29.7.0
```

I'm pretty sure this is what I'm running into as well, for a slightly less minimal example if it's useful then have a look at https://github.com/mscharley/ioc-deco/pull/63

Seems to only be an issue with TS ~5.2, 5.1 is unaffected. It's similarly happening across Node 16, 18, and 20 for me as well.

I'd also be interested to know why it seems like ts-jest is messing with `moduleResolution` and especially `module` settings - it can trigger very different behaviour in projects if it gets toggled between different values. In my project I have both `module` and `moduleResolution` set explicitly to `node16` but I managed to get the same error out of Jest:

```
  ● Test suite failed to run

    error TS5110: Option 'module' must be set to 'Node16' when option 'moduleResolution' is set to 'Node16'.
```

--- a/src/legacy/ts-jest-transformer.ts
+++ b/src/legacy/ts-jest-transformer.ts
@@ -189,31 +189,29 @@ export class TsJestTransformer implements SyncTransformer {
   ): Promise<TransformedSource> {
     this._logger.debug({ fileName: sourcePath, transformOptions }, 'processing', sourcePath)
 
-    return new Promise(async (resolve, reject) => {
-      const configs = this._configsFor(transformOptions)
-      const shouldStringifyContent = configs.shouldStringifyContent(sourcePath)
-      const babelJest = shouldStringifyContent ? undefined : configs.babelJestTransformer
-      let result: TransformedSource
-      const processWithTsResult = this.processWithTs(sourceText, sourcePath, transformOptions)
-      result = {
-        code: processWithTsResult.code,
-      }
-      if (processWithTsResult.diagnostics?.length) {
-        reject(configs.createTsError(processWithTsResult.diagnostics))
-      }
-      if (babelJest) {
-        this._logger.debug({ fileName: sourcePath }, 'calling babel-jest processor')
+    const configs = this._configsFor(transformOptions)
+    const shouldStringifyContent = configs.shouldStringifyContent(sourcePath)
+    const babelJest = shouldStringifyContent ? undefined : configs.babelJestTransformer
+    let result: TransformedSource
+    const processWithTsResult = this.processWithTs(sourceText, sourcePath, transformOptions)
+    result = {
+      code: processWithTsResult.code,
+    }
+    if (processWithTsResult.diagnostics?.length) {
+      throw configs.createTsError(processWithTsResult.diagnostics)
+    }
+    if (babelJest) {
+      this._logger.debug({ fileName: sourcePath }, 'calling babel-jest processor')
 
-        // do not instrument here, jest will do it anyway afterwards
-        result = await babelJest.processAsync(result.code, sourcePath, {
-          ...transformOptions,
-          instrument: false,
-        })
-      }
-      result = this.runTsJestHook(sourcePath, sourceText, transformOptions, result)
+      // do not instrument here, jest will do it anyway afterwards
+      result = await babelJest.processAsync(result.code, sourcePath, {
+        ...transformOptions,
+        instrument: false,
+      })
+    }
+    result = this.runTsJestHook(sourcePath, sourceText, transformOptions, result)
 
-      resolve(result)
-    })
+    return result
   }
 
   private processWithTs(

